### PR TITLE
[GraphQL Client] Rework the Page type and APIs that return Option<Page> 

### PR DESCRIPTION
**Pull request by stefan-mysten**
This PR simplifies the return type of several APIs from `Result<Option<Page>, Error>` to `Result<Page, Error>`. 
The Page type has now a `is_empty` function to check if there is no information returned in this page.

It should simplify a bit working with the results of these APIs.

### CODE REVIEW

File: crates/sui-graphql-client/src/lib.rs [hunk 1]
```diff
@@ -393,14 +405,14 @@ impl Client {
                     None,
                 ).await?;
 
-                if let Some(page) = response {
-                    for object in page.data {
-                        if let Some(coin) = Coin::try_from_object(&object) {
+                if !response.is_empty() {
```
Comment (amnn) [inline crates/sui-graphql-client/src/lib.rs:408]: the point of the simplification is that you no longer need this outer check. This code can be written as:

```rust
for object in response.data() {
    if let Some(coin) = Coin::try_from_object(object) {
      yield coin.into_owned();
    }
}

after = response.page_info.end_cursor;
if after.is_none() {
  break;
}
```

File: crates/sui-graphql-client/src/lib.rs [hunk 2]
```diff
@@ -93,17 +93,30 @@ pub struct Page<T> {
...
+    /// Check if the page has data.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    /// Internal function to create a page with no data.
+    fn new_empty() -> Self {
+        Self::new(PageInfo::default(), vec![])
+    }
 }
```
Comment (amnn) [inline crates/sui-graphql-client/src/lib.rs:120]: You could also `derive(Default)` on `Page` for this.

**Review by amnn — COMMENTED** [review]
Some small notes!

File: crates/sui-graphql-client/src/lib.rs (hunk 2 above)
Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs:120]: Thanks! I remember trying this and it was complaining about something :D

Comment (amnn) [reply crates/sui-graphql-client/src/lib.rs:120]: Ah that's a shame, thanks for checking. The issue is that `derive(Default)` introduces a `Default` constraint on all type parameters, even if it's not strictly necessary (e.g. the `Default` impl for `Vec<T>` does not require `Default` on the `T`).

**Pull request merged**